src/btree/bt_slvg.c

```diff
@@ -1802,13 +1803,30 @@ __slvg_row_build_internal(WT_SESSION_IMPL *session, uint32_t leaf_cnt, WT_STUFF
             WT_ERR(__slvg_ovfl_ref_all(session, trk));
         }
         ++ref;
+
+        /*
+         * !!!
+         * There's a risk the page we're building is too large for the cache. The right fix would be
+         * to write the keys out to an on-disk file and delay allocating the page image until we're
+         * ready to reconcile the new root page, and then read keys in from that backing file during
+         * the reconciliation of the root page. For now, make sure the eviction threads don't see us
+         * as a threat.
+         */
+        if (page->memory_footprint > 10 * WT_MEGABYTE) {
```

Comment:
You are right that I don't like the approach, but I can't think of something simple that has a similar effect. Thanks for opening WT-5447 to explore the correct solution.

I'm inclined to pick a smaller value here - 1MB seems more sensible than 10MB. I'd also like to note that in other cases we have done this a different way - allowing pages to grow without eviction getting involved by not queuing pages for urgent eviction if we know the eviction can't succeed (see https://github.com/wiredtiger/wiredtiger/blob/develop/src/include/btree.i#L1437). It would be better if something like that would work here. The solution you outline will make WiredTiger statistics look weird and make it difficult to track down where the memory allocations are coming from if there get to be enough that the system crashes with OOM.

Reply:
sorry for the rambly review - I'm happy to talk if I was too vague.

Reply:
> I'm inclined to pick a smaller value here - 1MB seems more sensible than 10MB. I'd also like to note that in other cases we have done this a different way - allowing pages to grow without eviction getting involved by not queuing pages for urgent eviction if we know the eviction can't succeed (see https://github.com/wiredtiger/wiredtiger/blob/develop/src/include/btree.i#L1437). It would be better if something like that would work here. The solution you outline will make WiredTiger statistics look weird and make it difficult to track down where the memory allocations are coming from if there get to be enough that the system crashes with OOM.

Agreed. The problem is the page we're building isn't part of any  tree, and so eviction will never find it, eviction will simply spin through all of the pages it can find, looking for something to evict that will help the situation. Or, in other words, salvage is temporarily making the cache appear smaller than it is.

Reply:
@agorrod, for clarity, when I say "salvage is temporarily making the cache appear smaller than it is", what's happening is salvage is calling `__wt_cache_page_inmem_incr()` and other cache size routines, which increment statistics like `WT_CACHE.{bytes_inmem,bytes_dirty_intl,bytes_internal}` (and the corresponding `WT_BTREE` statistics, but in a tree that eviction can't discover.